File: src/algorithm/leetcode/Problem309.java
package algorithm.leetcode;

public class Problem309 {

    public int maxProfit(int[] prices) {
        if (prices.length == 0)
            return 0;
        if (prices.length == 1) {
            return 0;
        }
        int len = prices.length;
        int[] buy = new int[len];
        int[] sell = new int[len];
        buy[0] = prices[0] * -1;
        buy[1] = prices[1] < prices[0] ? prices[1] * -1 : prices[0] * -1;
        sell[0] = 0;
        for (int i = 1; i < len; i++) {
            if (i > 1)
                buy[i] = Math.max(sell[i - 2] - prices[i], buy[i - 1]);
            sell[i] = Math.max(buy[i - 1] + prices[i], sell[i - 1]);
        }
        return sell[len - 1];
    }

}

File: src/algorithm_cpp/problem309.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
	if (prices.size() == 0) return 0;
	vector<vector<int>> dp(vector<int, 3>, prices.size());
	dp[0][0] = -1 * prices[0];
	dp[0][1] = 0;
	dp[0][2] = 0;
	for (int i = 1; i < prices.size(); i++) {
	    dp[i][0] = max(dp[i - 1][0], dp[i - 1][2] - prices[i]);
	    dp[i][1] = dp[i - 1][0] + prices[i];
	    dp[i][2] = max(dp[i - 1][1], dp[i - 1][2]);
	}
	return max(dp[i][1], dp[i][2]);
    }
};
